### PR TITLE
Fix exceptions' handling

### PR DESCRIPTION
**Pull request by AR-May**
Second PR concerning exceptions' handling. Addresses issues #27, #11, #9, #8.

Changes:
- Add functions for getting HTTP headers.
- Make download errors visible (including the TLS issue).
- Error 404 is now processed separately: for some of the cases, getting error 404 is a normal flow of script execution. 
- Fix extraction failure processing.

### CODE REVIEW

File: src/dotnet-install.sh [hunk 1]
```diff
@@ -713,44 +732,56 @@ download() {
...
     if [ -z "$out_path" ]; then
-        wget -q $wget_options -O - "$remote_path" || failed=true
+        wget -q $wget_options -O - "$remote_path_with_credential" || failed=true
     else
-        wget $wget_options -O "$out_path" "$remote_path" || failed=true
+        wget $wget_options -O "$out_path" "$remote_path_with_credential" || failed=true
     fi
     if [ "$failed" = true ]; then
-        say_verbose "Wget download failed"
+        local response=$(get_http_header_wget $remote_path_with_credential)
```
Comment (AR-May) [inline src/dotnet-install.sh:778]: I could not get both the HTTP response code and the output of the download together with one request safely (as they could both go to standard output and I consider parsing it in order to separate them too risky), so I decided to ask for a header and get the HTTP response code from there in case of failure. Likely, in most of the cases we will have the same response. I am open to converse about that decision.

File: src/dotnet-install.ps1 [hunk 2]
```diff
@@ -289,16 +289,34 @@ function GetHTTPResponse([Uri] $Uri)
...
+                    $ErrorMsg += "$Response"
                 }
-
                 throw $ErrorMsg
             }
 
             return $Response
+
+        }
+        catch {
```
Comment (vlada-shubina) [inline src/dotnet-install.ps1:303]: You might want to catch HTTP Request Exception here, not to rely on status code. Then Status Code will be returned as one of the properties.
It is possible to catch certain types of Exception in PowerShell: https://docs.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-exceptions?view=powershell-7.1#catching-typed-exceptions

File: src/dotnet-install.sh [hunk 3]
```diff
@@ -713,44 +732,56 @@ download() {
     return 0
 }
 
+# Note, function writes to global variables $http_code and $download_error_msg
```
Comment (vlada-shubina) [inline src/dotnet-install.sh:735]: nit: probably better wording could be: `updates/sets global variables $http_code and $download_error_msg`. `Note, fuction` seems obvious for me

File: src/dotnet-install.sh [hunk 4]
```diff
@@ -816,39 +847,76 @@ install_dotnet() {
     zip_path="$(mktemp "$temporary_file_template")"
     say_verbose "Zip path: $zip_path"
 
-    say "Downloading link: $download_link"
 
     # Failures are normal in the non-legacy case for ultimately legacy downloads.
     # Do not output to stderr, since output to stderr is considered an error.
+    say "Downloading link: $download_link"
+
+    # the download function will fill variables $http_code and $download_error_msg in case of failure.
```
Comment (vlada-shubina) [inline src/dotnet-install.sh:855]: will set is a better wording

File: src/dotnet-install.sh [hunk 5]
```diff
@@ -816,39 +847,76 @@ install_dotnet() {
...
+            fi
+            if [ "$legacy_path_http_code" != "404" ]; then
+                say_err "$legacy_path_download_error_msg"
+            fi
+        fi
         return 1
     fi
 
     say "Extracting zip from $download_link"
-    extract_dotnet_package "$zip_path" "$install_root"
```
Comment (vlada-shubina) [inline src/dotnet-install.sh:851]: Is this change relevant to PR?

File: src/dotnet-install.sh [hunk 6]
```diff
@@ -816,39 +847,76 @@ install_dotnet() {
...
 
     #  if the download fails, download the legacy_download_link
     if [ "$download_failed" = true ]; then
-        say "Cannot download: $download_link"
+        case $primary_path_http_code in
+        404)
+            say "The primary path $download_link is not available."
+            ;;
+        2*)
+            say "Cannot download via the primary path $download_link."
```
Comment (vlada-shubina) [inline src/dotnet-install.sh:867]: Can we reuse `primary_path_download_error_msg` here?  Seems logic is duplicated

**Review by vlada-shubina — COMMENTED** [review]
minor comments.
Duplicated logic should be avoided if possible
PS script: I would prefer to handle status code by reading the exception status code, and relying on message if possible (open for discussion here)

File: src/dotnet-install.ps1 [hunk 7]
```diff
@@ -289,16 +289,34 @@ function GetHTTPResponse([Uri] $Uri)
...
-                 # The feed credential is potentially sensitive info. Do not log FeedCredential to console output.
-                $ErrorMsg = "Failed to download $Uri."
+                # The feed credential is potentially sensitive info. Do not log FeedCredential to console output.
+                $ErrorMsg = ""
                 if ($Response -ne $null) {
-                    $ErrorMsg += "  $Response"
+                    $ErrorMsg += "$Response"
                 }
-
                 throw $ErrorMsg
```
Comment (nohwnd) [inline src/dotnet-install.ps1:297]: `throw ""` will just say "Exception: " on the screen. Maybe just log the first part of Uri and not the parameters.

File: src/dotnet-install.sh (hunk 5 above)
Comment (AR-May) [reply src/dotnet-install.sh:851]: Yes, if the script goes further it will end up with the "failed to install with an unknown error" message, which is confusing and wrong. The "extraction failed" message already explains the problem and we should finish on it. So, this change also concerns error messages.

File: src/dotnet-install.ps1 (hunk 7 above)
Comment (AR-May) [reply src/dotnet-install.ps1:297]: This throw will be always caught and not showed to the user, and the re-throwing will be performed with a different message (the initial one could be a part of it, but an empty message is not a problem in this case). We do not want to log all these errors to the screen where they are visible, it will be confusing because some 404 errors are expected in some cases and it is perfectly fine by design.

The only goal of throwing here is to deliver the error message further, actually. I considered writing to a global variable instead, actually, but finally left this.  The decision is debatable, sure.

File: src/dotnet-install.sh (hunk 6 above)
Comment (AR-May) [reply src/dotnet-install.sh:867]: Yes, I do not like ending up with duplicated code too. Messages are quite similar, but not the same. I may try to use a single message in these two cases and it will give me an opportunity to reuse, but that will mean I either skip part of the information in messages (which link I now download, primary or legacy), or I provide information that users do not need to see if everything goes fine (Is it wget or curl download).

Comment (AR-May) [reply src/dotnet-install.sh:867]: Well, I just got an idea, if I put the information about the link being privacy or legacy in the message
```
say "Downloading link: $download_link"
``` 
it will allow me to skip this information when I write the error message and remove this duplication. I suppose I will go for that.

File: src/dotnet-install.ps1 (hunk 2 above)
Comment (AR-May) [reply src/dotnet-install.ps1:303]: Nice idea, I suppose I'd like to specialize the catch here. But I am not sure it will immediately help to remove parsing.
First, Property Status Code for HttpRequestException is a 5.0 feature, not sure I can use it:
https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httprequestexception.statuscode?view=net-5.0#System_Net_Http_HttpRequestException_StatusCode
And secondly, we need also to have this property for the exception we throw explicitly like in row 298: `throw $ErrorMsg`. So, we either need to write a custom exception class and throw it or to find the existing one to create there.

Comment (vlada-shubina) [reply src/dotnet-install.ps1:303]: Sorry for confusion. Previous message intend to say `throw` instead of catch. You might want to use: https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpresponsemessage.ensuresuccessstatuscode?view=net-5.0 method to do it.
Having StatusCode property is unfortunate, probably then it is easier to keep it as is as anyway parsing is required. 

Is it an option to save the http code to variable to avoid parsing? (similar to bash script). In this method the status code is still available via property and putting it to string, and parsing again doesn't look as a good idea.  Should we consider updating the method just to return the response and do error handling outside?

File: src/dotnet-install.ps1 (hunk 7 above)
Comment (AR-May) [reply src/dotnet-install.ps1]: Fixed. Used the `Data` property of the exception to deliver the message, as discussed.

File: src/dotnet-install.ps1 (hunk 2 above)
Comment (AR-May) [reply src/dotnet-install.ps1]: Fixed. Used the `Data` property of the exception to preserve the status code, as discussed.

**Review by vlada-shubina — APPROVED** [review]
Like the exception approach 👍

File: src/dotnet-install.ps1 [hunk 8]
```diff
@@ -750,24 +772,59 @@ $ZipPath = [System.IO.Path]::combine([System.IO.Path]::GetTempPath(), [System.IO
...
+Say "Downloading primary link $DownloadLink"
 try {
     DownloadFile -Source $DownloadLink -OutPath $ZipPath
 }
 catch {
-    Say "Cannot download: $DownloadLink"
+    if ($PSItem.Exception.Data.Contains("StatusCode")) {
+        $PrimaryDownloadStatusCode = $PSItem.Exception.Data["StatusCode"]
+    }
+    if ($PSItem.Exception.Data.Contains("ErrorMessage")) {
```
Comment (nohwnd) [inline src/dotnet-install.ps1:790]:
```suggestion
    }
    
    if ($PSItem.Exception.Data.Contains("ErrorMessage")) {
```

Comment (nohwnd) [reply src/dotnet-install.ps1:790]: nit: Newline to separate the if from else?

File: src/dotnet-install.ps1 [hunk 9]
```diff
@@ -750,24 +772,59 @@ $ZipPath = [System.IO.Path]::combine([System.IO.Path]::GetTempPath(), [System.IO
...
+        Say "Downloading legacy link $DownloadLink"
         try {
             DownloadFile -Source $DownloadLink -OutPath $ZipPath
         }
         catch {
-            Say "Cannot download: $DownloadLink"
+            if ($PSItem.Exception.Data.Contains("StatusCode")) {
+                $LegacyDownloadStatusCode = $PSItem.Exception.Data["StatusCode"]
+            }
+            if ($PSItem.Exception.Data.Contains("ErrorMessage")) {
```
Comment (nohwnd) [inline src/dotnet-install.ps1:816]:
```suggestion
            }
            
            if ($PSItem.Exception.Data.Contains("ErrorMessage")) {
```

Nit: separate else and if by newline.

File: src/dotnet-install.ps1 [hunk 10]
```diff
@@ -778,7 +835,19 @@ catch {
 }
 
 if ($DownloadFailed) {
-    throw "Could not find/download: `"$assetName`" with version = $SpecificVersion`nRefer to: https://aka.ms/dotnet-os-lifecycle for information on .NET Core support"
+    if (($PrimaryDownloadStatusCode -eq "404") -and ((-not $LegacyDownloadLink) -or ($LegacyDownloadStatusCode -eq "404"))) {
```
Comment (nohwnd) [inline src/dotnet-install.ps1:838]: nit: you sometimes compare the status code as string and sometimes as number - line 822), is that intentional?

File: src/dotnet-install.sh (hunk 1 above)
Comment (bozturkMSFT) [reply src/dotnet-install.sh:778]: I am a bit worried about this. Under heavy load, for instance, server could process one of the requests, but not the other. Is there no way to get the content and the response code separately?

File: src/dotnet-install.ps1 [hunk 11]
```diff
@@ -750,24 +772,59 @@ $ZipPath = [System.IO.Path]::combine([System.IO.Path]::GetTempPath(), [System.IO
...
+        $PrimaryDownloadStatusCode = $PSItem.Exception.Data["StatusCode"]
+    }
+    if ($PSItem.Exception.Data.Contains("ErrorMessage")) {
+        $PrimaryDownloadFailedMsg = $PSItem.Exception.Data["ErrorMessage"]
+    } else {
+        $PrimaryDownloadFailedMsg = $PSItem.Exception.Message
+    }
+
+    if ($PrimaryDownloadStatusCode -eq 404) {
+        Say "The link $DownloadLink is not available."
```
Comment (bozturkMSFT) [inline src/dotnet-install.ps1:797]: I think "The resource at $DownloadLink is not available." would be more accurate.

**Review by bozturkMSFT — APPROVED** [review]

File: src/dotnet-install.sh (hunk 1 above)
Comment (AR-May) [reply src/dotnet-install.sh:778]: If it processes the real request, the second one will not be called. If the real one failed and the request for the header is successful this bit of information will not show. I have not found a way to separate these two outputs except making two separate requests. We still can get both of them in stdout and try to parse the result, but I do not like that idea more than the implemented one. I have not found a way to properly separate the content and the response code. If one suggests the way how to get them separately (when the content goes to stdout), I will be glad to consider that.

**Review by DavidKarlas — APPROVED** [review]
